### DISCUSSION TITLE
Adding Agents

### DISCUSSION THREAD

**Discussion by danielmiessler** [danielmiessler/Fabric#46] [Ideas]
One thing we're really looking to do is add agent functionality to Fabric.

Specifically, I want to have endpoints that point to things like plan_trip, for example, which actually runs a CrewAI team to do that task. 

We would have all the required keys in the `~/.config/fabric/.env` file so that it just works.

Thoughts?

**Comment by dheerapat** [comment]
something like #44 ?

**Comment by danielmiessler** [comment]
No, more like CrewAI, or Langraph, or Autogen.

**Reply by chrissonntag** [reply]
I've created a generic AutoGen setup today: https://github.com/chrissonntag/yaas

It's still in early days, but I'm thinking it could potentially be integrated into fabric. With this setup, we can configure workflows using simple MD files, making everything super easy to manage. The key to getting consistent results is giving clear and quality instructions. So if we manage to do so via appropriate patterns, surplus implement the one or another "skill", we should get it going 😁

**Comment by davejpeters** [comment]
Any updates on integration with crewAI?

**Reply by xssdoctor** [reply]
There's already an --agents flag

**Reply by davejpeters** [reply]
Is there a resource on how to use --agents?

**Reply by davejpeters** [reply]
Im exhausted. I’ve been up all night shaking with excitement. I’ve managed to install fabric and llama3:70b. I have a Windows 10 home laptop. Nothing special. I have so many questions. To start, how can I integrate my own prompts? I’m newish to python but want to learn what I need to to make this project more personalized.

**Reply by davejpeters** [reply]
I want to make a yt video on how others can do the same for very little cost. Is this a good idea?

**Reply by xssdoctor** [reply]
There are two ways to create your own patterns. You can add  folder to .config/python/patterns with your pattern name. In the folder put in a file called system.md with your custom pattern inside. Another way is in the gui. Just run fabric --gui and there is a prompt creator in there

**Reply by davejpeters** [reply]
Thank you! I will try this when I get home. Is there a way to add my own APIs? Are the |, &, && commands correct for creating flows?

**Reply by xssdoctor** [reply]
What do you mean add your own apis?

**Reply by davejpeters** [reply]
I'm looking into praisonAI and crewAI rn. Fabric integrates autoGen correct? I think I answered my own question because crew, praison, autogen offer this option of inserting APIs to call ( I think ). I'm still very new but learning fast and brainstorming.

**Reply by xssdoctor** [reply]
--agent uses praisonai. Just describe what you want to do and the agent will start working

**Reply by davejpeters** [reply]
I see praisonAI demos often using a SERP API key. How does fabric achieve web retrieval? Is there a way to define what tools are available to praisonAI and thus fabric?

**Reply by davejpeters** [reply]
Does this mean my praison agents will be able to be imported?? 🤞 I'm new to python but this is so cool i have to learn it

**Reply by davejpeters** [reply]
So if I understand the stack correctly - I would use LangGraph -> CrewAI -> PraisonAI -> fabric as the dev stack. Could I train models using this stack? Is it possible for a small team to specialize models for specific tasks to accomplish specific goals, possibly by renting GPUs? 

In theory, could I train many smaller models on Boolean questions then train more models with slightly more expertise on the same subject to work as managers. If so, I could also implement rag for each, possibly with regex. Thinking if an army of "Russian dolls" models trained for specific tasks, possibly led by hierarchical "need to know" managers.

**Reply by davejpeters** [reply]
Can fabric be used for training ?

**Reply by davejpeters** [reply]
Whoa 😳 I think I'm getting a better idea of what I'm working with, i.e. --agents 🤯

**Comment by davejpeters** [comment]
I think adding doraindarko/maestro would be a wonderful idea. Is there a way to include my own agents in fabric? I now have agents to do tasks and would like fabric to have the ability to call them like plan_trip.py. If I place an agent file in /agents, will that work?

**Comment by geektx** [comment]
> Does this mean my praison agents will be able to be imported?? 🤞 I'm new to python but this is so cool i have to learn it

Have you discovered a method to bring them in? I have created a pattern that creates agents.yaml playbooks that I really want to pipe to --agents.

**Reply by davejpeters** [reply]
I believe I have but have yet to test it

Might try tonight

**Comment by nic0711** [comment]
How about integrating something like this?

mendableai/firecrawl

**Comment by prabhic** [comment]
Today have tried 
**AGENT TEST**
` echo "generate qt c++ code to show 100 qml delegates horizontally. qml delegate contains image widget, and passe to c++ code. widget code should be in purely c++ without using any qt default listview framework. and should implement using qt quick framework without default listview. it should be optimized for memory and performance" | fabric --agents`

**OUTPUT**
Firstly I am Impressed to see that agents are created automatically. and the code is generated successfully.   Created following Agents. [DEBUG]: == Working Agent: QML Designer,  [DEBUG]: == Working Agent: C++ Developer,   [DEBUG]: == Working Agent: Performance Optimizer. and code is generated. 

**FEEDBACK**
Understandably, generated code have deviation from my expectation. Mostly it can be because of the clarity in my input prompt. 

**SUGGESTION**
However If we are able to iterate on agent description and role, and task, on command line itself, it will have huge uplift. 

Nice to see and experiment on fabric, great work.

**Reply by ajoslin103** [reply]
@prabhic are you using the Python or GO version of Fabric ?

**Reply by LaBalaDeOro** [reply]
Prolly Python version, since Go version wasn't released until later in the year.
